File: day_02/part_2/solution.js
const fs = require("fs");

let safeCount = 0;

const fileContent = fs.readFileSync("../input.txt", "utf8");
const lines = fileContent.trim().split("\n");

for (let line of lines) {
    const report = line.split(" ").map(Number);

    let increasing = true;
    let decreasing = true;
    let validDifferences = true;

    for (let i = 0; i < report.length - 1; i++) {
        if (report[i] > report[i + 1]) {
            increasing = false;
        }
        if (report[i] < report[i + 1]) {
            decreasing = false;
        }
        if (!(1 <= Math.abs(report[i] - report[i + 1]) && Math.abs(report[i] - report[i + 1]) <= 3)) {
            validDifferences = false;
        }
    }

    let isSafe = (increasing || decreasing) && validDifferences;

    if (!isSafe) {
        let isSafeAfterRemoval = false;

        for (let i = 0; i < report.length; i++) {
            const modifiedReport = [...report.slice(0, i), ...report.slice(i + 1)];

            increasing = true;
            decreasing = true;
            validDifferences = true;

            for (let j = 0; j < modifiedReport.length - 1; j++) {
                if (modifiedReport[j] > modifiedReport[j + 1]) {
                    increasing = false;
                }
                if (modifiedReport[j] < modifiedReport[j + 1]) {
                    decreasing = false;
                }
                if (!(1 <= Math.abs(modifiedReport[j] - modifiedReport[j + 1]) && Math.abs(modifiedReport[j] - modifiedReport[j + 1]) <= 3)) {
                    validDifferences = false;
                }
            }

            if ((increasing || decreasing) && validDifferences) {
                isSafeAfterRemoval = true;
                break;
            }
        }

        if (isSafeAfterRemoval) {
            safeCount += 1;
        }
    } else {
        safeCount += 1;
    }
}

console.log(`Answer: ${safeCount}`);

File: day_02/part_2/solution.py
safe_count = 0

with open("../input.txt", 'r') as file:
    for line in file:
        report = list(map(int, line.strip().split()))

        increasing = True
        decreasing = True
        valid_differences = True

        for i in range(len(report) - 1):
            if report[i] > report[i + 1]:
                increasing = False
            if report[i] < report[i + 1]:
                decreasing = False
            if not (1 <= abs(report[i] - report[i + 1]) <= 3):
                valid_differences = False

        is_safe = (increasing or decreasing) and valid_differences

        if not is_safe:
            is_safe_after_removal = False
            for i in range(len(report)):
                modified_report = report[:i] + report[i + 1:]

                increasing = True
                decreasing = True
                valid_differences = True

                for j in range(len(modified_report) - 1):
                    if modified_report[j] > modified_report[j + 1]:
                        increasing = False
                    if modified_report[j] < modified_report[j + 1]:
                        decreasing = False
                    if not (1 <= abs(modified_report[j] - modified_report[j + 1]) <= 3):
                        valid_differences = False

                if (increasing or decreasing) and valid_differences:
                    is_safe_after_removal = True
                    break

            if is_safe_after_removal:
                safe_count += 1
        else:
            safe_count += 1

print(f"Answer: {safe_count}")